### PR TITLE
Make flexible naming in explanation.save

### PR DESCRIPTION
**Pull request by GalyaZalesskaya**
Allows flexibly name the files with the prefix, suffix, and postfix. Add possibility to save confidence scores in saliency map name
 
        save(output_dir) -> aeroplane.jpg
        save(output_dir, prefix="image_name_target_") -> image_name_target_aeroplane.jpg
        save(output_dir, postfix="_class_map") -> aeroplane_class_map.jpg
        save(
            output_dir, prefix="image_name_", postfix="_conf_", confidence_scores=scores
        ) -> image_name_aeroplane_conf_0.85.jpg

### CODE REVIEW

**Comment by GalyaZalesskaya** [comment]
Probably, it's better to update the documentation according to use cases, but I'm not sure, which piece of documentation

File: docs/source/user-guide.md [hunk 1]
```diff
@@ -327,6 +331,78 @@ model_xai = xai.insert_xai(
...
+
+def preprocess_fn(image: np.ndarray) -> np.ndarray:
+    """Preprocess the input image."""
+    resized_image = cv2.resize(src=image, dsize=(224, 224))
+    expanded_image = np.expand_dims(resized_image, 0)
+    return expanded_image
+
+def postprocess_fn(output: OVDict):
+    """Postprocess the model output."""
+    return output["logits"]
```
Comment (goodsong81) [inline docs/source/user-guide.md]:
```suggestion
from from typing import Mapping
import openvino_xai as xai

def preprocess_fn(image: np.ndarray) -> np.ndarray:
    """Preprocess the input image."""
    resized_image = cv2.resize(src=image, dsize=(224, 224))
    expanded_image = np.expand_dims(resized_image, 0)
    return expanded_image

def postprocess_fn(output: Mapping):
    """Postprocess the model output."""
    return output["logits"]
```

Comment (goodsong81) [reply docs/source/user-guide.md]: Could you possibly replace the verbose `OVDict` import by `Mapping` in other code or docs?
Please refer to https://github.com/openvinotoolkit/openvino_xai/pull/45

File: openvino_xai/explainer/explanation.py [hunk 2]
```diff
@@ -132,21 +132,61 @@ def _select_target_indices(
...
         return target_indices
 
-    def save(self, dir_path: Path | str, name: str | None = None) -> None:
-        """Dumps saliency map."""
+    def save(
+        self,
+        dir_path: Path | str,
+        prefix_name: str | None = "",
+        suffix_name: str | None = "target",
+        postfix_name: str | None = "",
```
Comment (goodsong81) [inline openvino_xai/explainer/explanation.py]: Is it usual to put "suffix" and "postfix" together?
I thought they are rather same things. Anyway, isn't it a little bit confusing to users?

File: docs/source/user-guide.md [hunk 3]
```diff
@@ -327,6 +331,78 @@ model_xai = xai.insert_xai(
...
+explanation.save(OUTPUT_PATH, "image_name")  # image_name_target_aeroplane.jpg
+explanation.save(OUTPUT_PATH, prefix_name="image_name")  # image_name_target_aeroplane.jpg
+explanation.save(OUTPUT_PATH)  # target_aeroplane.jpg
+# Avoid "target" in names
+explanation.save(OUTPUT_PATH, prefix_name="image_name", target_suffix="")  # image_name_aeroplane.jpg
+explanation.save(OUTPUT_PATH, target_suffix="", postfix_name="class")  # aeroplane_class.jpg
+explanation.save(OUTPUT_PATH, target_suffix="")  # aeroplane.jpg
+
+# Save saliency maps with confidence scores
+explanation.save(OUTPUT_PATH, postfix_name="conf", confidence_scores=scores_dict)  # target_aeroplane_conf_0.92.jpg```
```
Comment (goodsong81) [inline docs/source/user-guide.md]: This example is inconsistent with the actual parameter definition.

**Review by goodsong81 — CHANGES_REQUESTED** [review]
Thank you for the usability enhancement!
BTW, I feel a bit confused by the terms. Can we refine?
Anyway, we need match the code & user guide.

File: openvino_xai/explainer/explanation.py (hunk 2 above)
Comment (GalyaZalesskaya) [reply openvino_xai/explainer/explanation.py]: `Suffix_name` goes before inserted label_name, while the `postfix` - after. I will add the doc comment like this.

        For the name 'image_name_target_aeroplane.jpg': prefix = 'image_name', suffix = 'target', label name = 'aeroplane', postfix = ''.

Theoretically, we can combine `prefix` and `suffix` since they will be combined and put before `label_name`, but they have different meanings. `Suffix` -  to name the `target` part, where we will highlight the label name of saliency map.

I'm looking for a way to make it more clear

File: docs/source/user-guide.md (hunk 3 above)
Comment (GalyaZalesskaya) [reply docs/source/user-guide.md]: Thank you for noticing

File: openvino_xai/explainer/explanation.py (hunk 2 above)
Comment (goodsong81) [reply openvino_xai/explainer/explanation.py]: What about
prefix_name -> file_prefix
suffix_name -> target_prefix
postfix_name -> target_suffix

**Comment by negvet** [comment]
I am sorry, maybe I misunderstood it. But I am not sure why we need three string arguments. Can we can have just two?

We want the name to contain some target reference, this is for sure. This might be an index or a label name, it is known. Let's refer to it as `target_id`.
Then, we want to let the user to add some info before and after `target_id`, it can be image name, test number or whatever.

So here is the structure with two optional string arguments:
`prefix` + `target_id` + `postfix`

Having `target_id=aeroplane`, we can get:
- `aeroplane`
- `image1_target_aeroplane` (`prefix=image1_target_`)
- `aeroplane_test2` (`postfix =_test2`)

Now, we also want to support `confidence_scores=scores`, which is "" by default, then we have the following name components:
`prefix` + `target_id` + `postfix` + `confidence`

Then we get:
- `image1_target_aeroplane_conf_0.8` (`prefix="image1_target_"`, `postfix="_conf_"`, `confidence_scores=scores`)

**Comment by GalyaZalesskaya** [comment]
@negvet I didn't like the idea of combining `image_name_prefix` and `target_prefix` since they have different meaning. One for image name, another - for target name. But the structure with 3 components indeed looks too complex, I agree.

My concern here, is that we put the responsibility to add `target_` to prefix to  the user, since there's no way for us to make it the default value. But probably, we don't need this `target_` at all, since the saliency map name will be well-read even without it.

**Comment by negvet** [comment]
> @negvet I didn't like the idea of combining `image_name_prefix` and `target_prefix` since they have different meaning. One for image name, another - for target name. But the structure with 3 components indeed looks too complex, I agree.
> 
> My concern here, is that we put the responsibility to add `target_` to prefix to the user, since there's no way for us to make it the default value. But probably, we don't need this `target_` at all, since the saliency map name will be well-read even without it.

In general case, there might be image name, some explain parameters, shooting conditions, timeframe, etc.
Let's keep it simple, without any assumptions about meanings, such as string should start with image name. Why not just let the user define what she want to use before and after `targer_id`? For this we need to let the user control prefix and postfix.

In our examples, we might use `image_name_target_aeroplane.jpg` convention, but user is not required to follow it.

**Review by goodsong81 — DISMISSED** [review]
Thanks for the update. I'm all OK for 2 args or 3 args.

**Review by negvet — APPROVED** [review]
Looks good, thanks!
